File: frontend/src/lib/projectCapabilities.ts
// ЕДИНСТВЕННАЯ точка правды о локальности проекта и доступности подсистем (ADR-016 §4).
// Источник истины — матрица `Project.capabilities`, которую бэк собирает из
// ProjectCapabilities.cs (DTO). Панели спрашивают «есть ли мой ключ в features и
// доступна ли группа», а НЕ проверяют deviceId руками.
//
// Сторож G10 (ProjectCapabilitiesGuardTests) разрешает обращения к project.deviceId
// ТОЛЬКО в этом файле — любая такая проверка в другом месте краснеет. Разрешённые
// формы здесь — через тернарник и условный оператор (без сравнения с null/undefined и
// без double-bang). Проверки на «локальность» живут в `isLocalProject`, остальной код
// берёт матрицу через getProjectCapabilities.

import { useMemo } from 'react';
import type {
  Project,
  ProjectCapabilitiesView,
  ProjectFeatureKey,
  ProjectCapabilityGroup,
} from '../types';
import { ProjectFeature } from '../types';
import { agentSupports, relaySupports, type ProjectRoute } from './deviceAgentRoutes';

// Состав групп возможностей — зеркало ProjectFeatures.FileBound/Platform/ServerContent
// на бэке. Менять синхронно с ProjectFeatures.cs при расширении.
const FILE_FEATURES: ProjectFeatureKey[] = [
  ProjectFeature.Files,
  ProjectFeature.Diff,
  ProjectFeature.Git,
  ProjectFeature.FileWatcher,
  ProjectFeature.Terminal,
  ProjectFeature.DevServers,
  ProjectFeature.Skills,
  ProjectFeature.Attachments,
];
const PLATFORM_FEATURES: ProjectFeatureKey[] = [
  ProjectFeature.Chat,
  ProjectFeature.History,
  ProjectFeature.Tasks,
  ProjectFeature.Memory,
  ProjectFeature.Personas,
  ProjectFeature.Notes,
  ProjectFeature.Costs,
  ProjectFeature.Tts,
];
const SERVER_CONTENT_FEATURES: ProjectFeatureKey[] = [
  ProjectFeature.Knowledge,
  ProjectFeature.CodeGraph,
  ProjectFeature.Dossiers,
  ProjectFeature.Docs,
  ProjectFeature.MapHygiene,
];
const TRANSCRIPT_FEATURES: ProjectFeatureKey[] = [
  ProjectFeature.LiveSubagents,
  ProjectFeature.WorkflowView,
  ProjectFeature.ChatBranch,
];

// Серверный проект без capabilities (старый бэк): всё доступно, матрица пустая.
// Это ЕДИНСТВЕННОЕ место, где мы «придумываем» матрицу; остальной код видит то,
// что вернёт getProjectCapabilities, и больше про deviceId не спрашивает.
// Группа «всё доступно, на сервере» — дефолт для проекта без capabilities (старый бэк
// или редкий код-путь) и для серверных проектов с заполненной матрицей (в этом случае
// бэк сам выставляет host='server'/available=true).
function makeServerGroup(features: ProjectFeatureKey[]): ProjectCapabilityGroup {
  return { host: 'server', available: true, reason: null, features };
}

// Возвращает матрицу возможностей: готовый `project.capabilities` или вырожденный
// «всё на сервере» для совместимости со старым бэком. Никаких выводов про deviceId —
// только матрица.
export function getProjectCapabilities(project: Project | null | undefined): ProjectCapabilitiesView {
  if (project?.capabilities) {
    // Бэк без группы транскрипта — серверный транскрипт, как было до неё
    return project.capabilities.transcript
      ? project.capabilities
      : { ...project.capabilities, transcript: makeServerGroup(TRANSCRIPT_FEATURES) };
  }
  // Совместимость со старым бэком: трактуем проект как серверный с полным набором.
  return {
    host: 'server',
    deviceId: null,
    files: makeServerGroup(FILE_FEATURES),
    platform: makeServerGroup(PLATFORM_FEATURES),
    serverContent: makeServerGroup(SERVER_CONTENT_FEATURES),
    transcript: makeServerGroup(TRANSCRIPT_FEATURES),
    exec: { available: true, reason: null },
  };
}

// В какой группе живёт ключ. Возвращает группу или null для незнакомого ключа.
// Сторож G10 разрешает в этом файле ВСЕ формы обращения к deviceId — потому проверки
// локальности тут, а в панелях только host/features/reason из матрицы.
function findGroup(cap: ProjectCapabilitiesView, feature: ProjectFeatureKey): ProjectCapabilityGroup | null {
  if (FILE_FEATURES.includes(feature)) return cap.files;
  if (PLATFORM_FEATURES.includes(feature)) return cap.platform;
  if (SERVER_CONTENT_FEATURES.includes(feature)) return cap.serverContent;
  if (TRANSCRIPT_FEATURES.includes(feature)) return cap.transcript ?? makeServerGroup(TRANSCRIPT_FEATURES);
  return null;
}

// Доступна ли возможность прямо сейчас. Панель решает «рисовать / показать причину»
// по этой функции. Не знаем ключа — false (защита от тихих регрессий при расширении
// списка ключей на бэке).
export function isFeatureAvailable(project: Project | null | undefined, feature: ProjectFeatureKey): boolean {
  const cap = getProjectCapabilities(project);
  const group = findGroup(cap, feature);
  if (!group) return false;
  return group.available && group.features.includes(feature);
}

// Готовый текст причины недоступности (null если доступно). UI показывает его как
// tooltip к скрытой панели/кнопке.
export function featureReason(project: Project | null | undefined, feature: ProjectFeatureKey): string | null {
  const cap = getProjectCapabilities(project);
  const group = findGroup(cap, feature);
  if (!group) return 'Возможность недоступна в этой версии';
  if (!group.features.includes(feature)) return null; // ключ не в этой группе — не показываем
  if (!group.available) return group.reason;
  return null;
}

// Куда ходят запросы файлов, git, сервисов и навыков проекта: на сервер, в агента устройства на этой
// машине или через ретранслятор сервера (проект открыт с другого устройства, только чтение).
// ЕДИНСТВЕННОЕ место этого решения — api.ts берёт его отсюда через deviceAgent.ts.
export type ProjectFilesRoute = 'server' | 'agent' | 'relay';

// Где браузер относительно машины проекта. Узнать это можно только у агента: here — агент на этом
// компьютере ответил; elsewhere — не ответил или это агент другого устройства; unknown — ещё не
// спрашивали (тогда сначала пробуем агента). Помнит ответ deviceAgent.ts
export type DevicePresence = 'unknown' | 'here' | 'elsewhere';

export function projectFilesRoute(project: Project | null | undefined, presence: DevicePresence = 'unknown'): ProjectFilesRoute {
  if (getProjectCapabilities(project).host !== 'device') return 'server';
  return presence === 'elsewhere' ? 'relay' : 'agent';
}

// Есть ли действие на маршруте: у сервера — все, у агента — его список, у ретранслятора — только
// чтение из RELAY_ROUTES (deviceAgentRoutes.ts). Кнопку без маршрута панель прячет — так и
// получается «ни одного контрола записи» с другого устройства без проверок в самих панелях
export function routeSupports(filesRoute: ProjectFilesRoute, route: ProjectRoute): boolean {
  if (filesRoute === 'server') return true;
  return filesRoute === 'relay' ? relaySupports(route) : agentSupports(route);
}

export function projectSupportsRoute(project: Project | null | undefined, route: ProjectRoute, presence: DevicePresence = 'unknown'): boolean {
  return routeSupports(projectFilesRoute(project, presence), route);
}

// Почему действия нет у проекта (null — есть). Текст — для title недоступной кнопки.
const ROUTE_REASONS: Partial<Record<ProjectRoute, string>> = {
  'POST preview/external-link': 'Доступ снаружи идёт через поддомен сервера, а сервис локального проекта работает на его компьютере',
};
export const DEVICE_ROUTE_REASON = 'У локального проекта это действие недоступно: агент устройства его не умеет';
export const RELAY_ROUTE_REASON = 'С другого устройства локальный проект открыт только для просмотра: изменить его можно на компьютере проекта';

export function projectRouteReason(project: Project | null | undefined, route: ProjectRoute, presence: DevicePresence = 'unknown'): string | null {
  const filesRoute = projectFilesRoute(project, presence);
  if (routeSupports(filesRoute, route)) return null;
  if (filesRoute === 'relay') return RELAY_ROUTE_REASON;
  return ROUTE_REASONS[route] ?? DEVICE_ROUTE_REASON;
}

// Где работает ключ — сервер/устройство/выключено. Полезно для бейджей («на устройстве»).
export function hostFor(project: Project | null | undefined, feature: ProjectFeatureKey): 'server' | 'device' | 'off' {
  const cap = getProjectCapabilities(project);
  const group = findGroup(cap, feature);
  if (!group) return 'off';
  return group.host;
}

// Можно ли запускать ход прямо сейчас (chat send). reason — готовый текст «Устройство
// офлайн» и т.п., который UI кладёт в баннер отправки и в карточку ошибки.
export function canRunTurn(project: Project | null | undefined): { available: boolean; reason: string | null } {
  const cap = getProjectCapabilities(project);
  return cap.exec;
}

// Это ЛОКАЛЬНЫЙ проект (привязан к устройству). Единственная функция с этим именем
// во всём фронте; через неё идёт любой показ «локальности» (бейдж, заголовок,
// маршрутизация ассистента). Вне этого файла её не зовут.
export function isLocalProject(project: Project | null | undefined): boolean {
  // Прямое обращение к deviceId допустимо ТОЛЬКО в этом файле (сторож G10). Условие
  // без сравнения с null и без double-bang шаблон не ловит — ловит только формы,
  // указанные в самом тесте сторожа.
  return project?.deviceId ? true : false;
}

// Хук для компонентов: стабильная обёртка над getProjectCapabilities. Сейчас без
// подписки — матрица приходит с проектом в REST-ответе и обновляется целиком при
// refresh. Если позже понадобится реактивный стор (событие device_online_changed),
// точка расширения здесь, не в компонентах.
export function useProjectCapabilities(project: Project | null | undefined): ProjectCapabilitiesView {
  return useMemo(() => getProjectCapabilities(project), [project?.capabilities, project?.deviceId]);
}

// Удобный хук-чекер возможности — для частого использования в JSX
// (useProjectFeature(project, 'files') вместо isFeatureAvailable(project, 'files')).
export function useProjectFeature(project: Project | null | undefined, feature: ProjectFeatureKey): boolean {
  return useMemo(() => isFeatureAvailable(project, feature), [project?.capabilities, project?.deviceId, feature]);
}

// Бейдж «устройство офлайн» / «не готово»: project.device?.online=false или
// harnessReady=false. Возвращает короткую строку для UI; null если всё хорошо.
// Зовётся в шапке чата и в диалогах — НЕ в панелях (у панелей свои причины через
// featureReason).
export function deviceOfflineLabel(project: Project | null | undefined): string | null {
  if (!project?.device) return null;
  if (!project.device.online) return 'Устройство офлайн';
  if (!project.device.harnessReady) {
    return project.device.harnessProblem ?? 'Агент устройства не готов';
  }
  return null;
}

// Бейдж локального проекта для шапки чата: «на устройстве · имя · офлайн». null — проект
// серверный. offline — устройство не в сети или агент не готов: тогда бейдж предупреждает,
// а в title полная причина из deviceOfflineLabel. short — для узкой шапки, где длинный
// текст обрезался бы как раз на состоянии
export function projectDeviceBadge(project: Project | null | undefined): { text: string; short: string; offline: boolean; title: string } | null {
  if (!isLocalProject(project)) return null;
  const problem = deviceOfflineLabel(project);
  const name = project?.device?.name;
  const state = !problem ? null : project?.device?.online === false ? 'офлайн' : 'не готово';
  const where = `Проект живёт на устройстве${name ? ` «${name}»` : ''}`;
  return {
    text: ['на устройстве', name, state].filter(Boolean).join(' · '),
    short: state ? `устройство ${state}` : 'на устройстве',
    offline: problem !== null,
    title: problem ? `${where}. ${problem}` : where,
  };
}

File: backend/ClaudeHomeServer.Core/Models/ProjectCapabilities.cs
using ClaudeHomeServer.Protocol;
using ClaudeHomeServer.Services;
using ClaudeHomeServer.Services.Execution;

namespace ClaudeHomeServer.Models;

/// <summary>
/// Где работает группа подсистем проекта (ADR-016 §4): на сервере, на устройстве (в агенте)
/// или нигде. Строки, а не enum: в DTO едут как есть, фронт сравнивает их литералами.
/// </summary>
public static class CapabilityHost
{
    public const string Server = "server";
    public const string Device = "device";
    public const string Off = "off";
}

/// <summary>Состояние одной группы подсистем у конкретного проекта.</summary>
/// <param name="Host">Где группа работает — <see cref="CapabilityHost"/>.</param>
/// <param name="Available">Можно ли пользоваться группой прямо сейчас.</param>
/// <param name="Reason">Почему недоступна — готовый текст для человека; null, если доступна.</param>
/// <param name="Features">Состав группы — ключи <see cref="ProjectFeatures"/>.</param>
public sealed record ProjectCapabilityGroup(string Host, bool Available, string? Reason, IReadOnlyList<string> Features);

/// <summary>Можно ли запускать ход проекта прямо сейчас.</summary>
public sealed record ProjectExecCapability(bool Available, string? Reason);

/// <summary>Вердикт для фоновой работы по проекту (ADR-016, план §5).</summary>
public enum ProjectBackgroundVerdict
{
    /// <summary>Запускать можно: серверный проект или готовое устройство.</summary>
    Ready,
    /// <summary>Устройство не готово (офлайн, нет exec, харнес) — ждать его выхода в онлайн.</summary>
    WaitDevice,
    /// <summary>Устройства нет или оно отозвано — ждать нечего.</summary>
    DeviceGone,
}

/// <summary>
/// Вердикт фоновой работы с устройством проекта. <see cref="DeviceId"/> — устройство
/// локального проекта (null у серверного), <see cref="Reason"/> — текст для человека.
/// </summary>
public sealed record ProjectBackgroundGate(ProjectBackgroundVerdict Verdict, string? DeviceId, string? Reason)
{
    public static readonly ProjectBackgroundGate Ready = new(ProjectBackgroundVerdict.Ready, null, null);

    public bool IsReady => Verdict == ProjectBackgroundVerdict.Ready;
    public bool MustWait => Verdict == ProjectBackgroundVerdict.WaitDevice;
}

/// <summary>
/// Ключи подсистем проекта для матрицы (ADR-016 §4). Фронт скрывает панели по группе, в
/// которую входит ключ, а не по <c>deviceId</c> проекта.
/// </summary>
public static class ProjectFeatures
{
    // Привязаны к файлам проекта
    public const string Files = "files";
    public const string Diff = "diff";
    public const string Git = "git";
    public const string FileWatcher = "fileWatcher";
    public const string Terminal = "terminal";
    public const string DevServers = "devServers";
    public const string Skills = "skills";
    public const string Attachments = "attachments";

    // Платформа
    public const string Chat = "chat";
    public const string History = "history";
    public const string Tasks = "tasks";
    public const string Memory = "memory";
    public const string Personas = "personas";
    public const string Notes = "notes";
    public const string Costs = "costs";
    public const string Tts = "tts";

    // Нужен контент проекта на сервере
    public const string Knowledge = "knowledge";
    public const string CodeGraph = "codeGraph";
    public const string Dossiers = "dossiers";
    public const string Docs = "docs";
    public const string MapHygiene = "mapHygiene";

    // Нужен транскрипт CLI на сервере (ADR-016: у локального проекта он живёт на устройстве)
    public const string LiveSubagents = "liveSubagents";
    public const string WorkflowView = "workflowView";
    public const string ChatBranch = "chatBranch";

    public static readonly IReadOnlyList<string> FileBound =
        [Files, Diff, Git, FileWatcher, Terminal, DevServers, Skills, Attachments];

    public static readonly IReadOnlyList<string> Platform =
        [Chat, History, Tasks, Memory, Personas, Notes, Costs, Tts];

    public static readonly IReadOnlyList<string> ServerContent =
        [Knowledge, CodeGraph, Dossiers, Docs, MapHygiene];

    public static readonly IReadOnlyList<string> Transcript =
        [LiveSubagents, WorkflowView, ChatBranch];
}

/// <summary>
/// Матрица возможностей проекта (ADR-016 §4) — ЕДИНСТВЕННАЯ точка правды о том, серверный
/// проект или локальный и что из этого следует, для фронта и серверных guard'ов. Инлайновых
/// проверок локальности (<c>DeviceId != null</c>, <c>IsLocal</c>) вне этого файла не заводим:
/// их ловит сторож G10 (<c>ProjectCapabilitiesGuardTests</c>).
///
/// Четыре группы: <see cref="Files"/> — подсистемы, привязанные к файлам (у локального проекта —
/// в агенте устройства), <see cref="Platform"/> — всегда на сервере, <see cref="ServerContent"/>
/// — нужен контент проекта на сервере (у локального выключены), <see cref="Transcript"/> —
/// механики, читающие транскрипт CLI с диска сервера (у локального выключены).
/// </summary>
public sealed record ProjectCapabilities(
    string Host,
    string? DeviceId,
    ProjectCapabilityGroup Files,
    ProjectCapabilityGroup Platform,
    ProjectCapabilityGroup ServerContent,
    ProjectCapabilityGroup Transcript,
    ProjectExecCapability Exec)
{
    public const string DeviceMissingReason = "Устройство проекта не найдено или отозвано";
    public const string DeviceOfflineReason = "Устройство проекта не в сети";
    public const string NoExecReason = "На устройстве нет агента AI Home для локальных проектов: чаты проекта не могут работать";
    public const string NoFilesReason = "Агент устройства пока не открывает файлы проекта: обновите агента";
    public const string NoRelayReason = "Агент устройства не умеет показывать файлы проекта на других устройствах: обновите агента";
    public const string NotDeviceBoundReason = "Проект не локальный: его файлы на сервере";
    public const string ServerContentOffReason = "У локального проекта недоступно: его файлы лежат на устройстве, а не на сервере";
    public const string TranscriptOnDeviceReason =
        "У локального проекта недоступно: подробная история чата хранится на его устройстве";

    /// <summary>Проект привязан к устройству. Единственная проверка локальности во всём коде.</summary>
    public static bool IsDeviceBound(Project project) => project.DeviceId is not null;

    /// <summary>Файлы проекта лежат на диске сервера — вход серверной файловой подсистемы допустим.</summary>
    public static bool FilesOnServer(Project project) => !IsDeviceBound(project);

    /// <summary>Работает ли у проекта группа «нужен контент на сервере» (Dify, CodeGraph, досье, Docs, уборка карты).</summary>
    public static bool ServerContentEnabled(Project project) => !IsDeviceBound(project);

    /// <summary>
    /// Транскрипты CLI (<c>.jsonl</c>) чатов проекта лежат на диске сервера. У локального проекта
    /// они живут только на устройстве: сервер их не ищет, не копирует и не удаляет — промах
    /// поиска по своему пути он принял бы за «транскрипта нет».
    /// </summary>
    public static bool TranscriptOnServer(Project project) => !IsDeviceBound(project);

    /// <summary>
    /// Ключ папки проекта — пара «устройство + путь» (ADR-016 §1): одна и та же строка пути на
    /// сервере и на устройстве даёт разные ключи. У серверного проекта ключ совпадает с
    /// прежним <see cref="PathNormalizer.NormalizePath"/> — сторы на диске не мигрируют.
    /// </summary>
    public static string FolderKey(Project project) => FolderKey(project.DeviceId, project.RootPath);

    public static string FolderKey(string? deviceId, string rootPath) =>
        deviceId is null
            ? PathNormalizer.NormalizePath(rootPath)
            // Путь чужой машины: GetFullPath сервера его исказил бы (Windows-путь на Linux),
            // поэтому только разделители, хвост и регистр
            : "device:" + deviceId + ":" + NormalizeDevicePath(rootPath).ToLowerInvariant();

    /// <summary>
    /// Ключ датасета знаний проекта. null — у проекта нет серверного датасета вовсе (локальный:
    /// группа «контент на сервере» выключена), поэтому одноимённая серверная папка не делит с
    /// ним базу знаний.
    /// </summary>
    public static string? KnowledgeRoot(Project project) =>
        ServerContentEnabled(project) ? project.RootPath : null;

    /// <summary>
    /// Путь на устройстве в каноничной форме: абсолютный (корень диска Windows или «/»), без
    /// сегментов «.»/«..», единый разделитель по виду пути, без хвостового разделителя. На
    /// сервере не существует — проверять его наличие некому, кроме агента.
    /// </summary>
    public static string NormalizeDevicePath(string rootPath)
    {
        var raw = rootPath.Trim();
        var windows = raw.Length >= 3 && char.IsAsciiLetter(raw[0]) && raw[1] == ':' && raw[2] is '\\' or '/';
        if (!windows && !raw.StartsWith('/'))
            throw new ArgumentException("Путь на устройстве должен быть абсолютным: «C:\\папка» или «/папка»");
        var sep = windows ? '\\' : '/';
        var segments = raw.Split(['\\', '/'], StringSplitOptions.RemoveEmptyEntries);
        if (segments.Any(s => s is "." or ".."))
            throw new ArgumentException("Путь на устройстве не должен содержать «.» и «..»");
        return windows
            ? segments.Length == 1 ? segments[0] + sep : string.Join(sep, segments)
            : "/" + string.Join(sep, segments);
    }

    /// <summary>
    /// Можно ли привязать проект к устройству: флаг <c>local-projects</c> владельца и возможность
    /// <c>exec</c> у устройства. null — можно, иначе готовый текст отказа. Онлайн не требуется:
    /// возможность хранится у устройства и в офлайне.
    /// </summary>
    public static string? BindRefusal(bool localProjectsEnabled, DeviceExecStatus? device)
    {
        if (!localProjectsEnabled) return "Локальные проекты выключены (экспериментальная функция «Локальные проекты»)";
        if (device is null) return DeviceMissingReason;
        if (!device.HasCapability(DeviceCapabilities.Exec)) return NoExecReason;
        return null;
    }

    /// <summary>
    /// Потолок ожидания устройства для разовой фоновой работы (ADR-016, вариант А плана §5):
    /// то же окно, что у автозапуска задач по сроку. Истёк — отказ с уведомлением, а не тишина.
    /// </summary>
    public static readonly TimeSpan DeviceWaitCeiling = TimeSpan.FromHours(24);

    /// <summary>
    /// Можно ли фоновой работе (исполнитель задачи, волна штаба, очередь чата, автоматизация,
    /// опрос сторожа) запускать ход проекта прямо сейчас. Серверный проект — всегда да.
    /// Локальный: устройство готово — да; устройства нет (отозвано) — ждать нечего, отказ;
    /// иначе (офлайн, нет exec, харнес не готов) — ждать выхода устройства в онлайн.
    /// </summary>
    public static ProjectBackgroundGate BackgroundGate(Project project, DeviceExecStatus? device)
    {
        if (!IsDeviceBound(project)) return ProjectBackgroundGate.Ready;
        if (device is null)
            return new ProjectBackgroundGate(ProjectBackgroundVerdict.DeviceGone, project.DeviceId, DeviceMissingReason);
        var exec = For(project, device).Exec;
        return exec.Available
            ? new ProjectBackgroundGate(ProjectBackgroundVerdict.Ready, project.DeviceId, null)
            : new ProjectBackgroundGate(ProjectBackgroundVerdict.WaitDevice, project.DeviceId, exec.Reason);
    }

    /// <summary>
    /// Чтение файлов проекта с другого устройства через ретранслятор (ADR-016 §5): только у
    /// локального проекта и только пока его устройство онлайн и объявило ретранслятор.
    /// null — можно, иначе причина для человека.
    /// </summary>
    public static string? RelayRefusal(Project project, DeviceExecStatus? device) =>
        !IsDeviceBound(project) ? NotDeviceBoundReason
        : device is null ? DeviceMissingReason
        : !device.Online ? DeviceOfflineReason
        : !device.HasCapability(DeviceCapabilities.Relay) ? NoRelayReason
        : null;

    /// <summary>
    /// Матрица для проекта. <paramref name="device"/> — состояние устройства проекта из шва
    /// <see cref="IDeviceExecChannel"/>; у серверного проекта не нужен, у локального null
    /// означает «устройства нет» (отозвано или канал недоступен).
    /// </summary>
    public static ProjectCapabilities For(Project project, DeviceExecStatus? device)
    {
        var platform = new ProjectCapabilityGroup(CapabilityHost.Server, true, null, ProjectFeatures.Platform);
        if (!IsDeviceBound(project))
            return new ProjectCapabilities(
                CapabilityHost.Server, null,
                new ProjectCapabilityGroup(CapabilityHost.Server, true, null, ProjectFeatures.FileBound),
                platform,
                new ProjectCapabilityGroup(CapabilityHost.Server, true, null, ProjectFeatures.ServerContent),
                new ProjectCapabilityGroup(CapabilityHost.Server, true, null, ProjectFeatures.Transcript),
                new ProjectExecCapability(true, null));

        string? filesReason =
            device is null ? DeviceMissingReason
            : !device.Online ? DeviceOfflineReason
            : !device.HasCapability(DeviceCapabilities.Files) ? NoFilesReason
            : null;

        string? execReason =
            device is null ? DeviceMissingReason
            : !device.Online ? DeviceOfflineReason
            : !device.HasCapability(DeviceCapabilities.Exec) ? NoExecReason
            : !device.HarnessReady ? device.HarnessProblem ?? "Харнес устройства не готов"
            : null;

        return new ProjectCapabilities(
            CapabilityHost.Device, project.DeviceId,
            new ProjectCapabilityGroup(CapabilityHost.Device, filesReason is null, filesReason, ProjectFeatures.FileBound),
            platform,
            new ProjectCapabilityGroup(CapabilityHost.Off, false, ServerContentOffReason, ProjectFeatures.ServerContent),
            new ProjectCapabilityGroup(CapabilityHost.Off, false, TranscriptOnDeviceReason, ProjectFeatures.Transcript),
            new ProjectExecCapability(execReason is null, execReason));
    }
}
